### ISSUE TITLE
Eliminate the need for the LclJSONSerialization code 

### ISSUE THREAD

**Issue by rolivieri** [IBM-Swift/Kitura#362]
Eliminate the need for the LclJSONSerialization code and instead use the implementation of NSJSONSerialization in Foundation. As a starting point, we'd need the following code to work on Linux just like it does today on OS X:

```
guard NSJSONSerialization.isValidJSONObject(self.object) else {
    throw NSError(domain: ErrorDomain, code: ErrorInvalidJSON, userInfo: [NSLocalizedDescriptionKey: "JSON is invalid"])
 }
return try NSJSONSerialization.data(withJSONObject: self.object, options: opt)
```

Ian has done the implementation of NSJSONSerialization, and it should now have been merged. Let's double check this is the case, test it out to verify it produces the expected outcome, and if so we can then remove the LclJSONSerialization code.

**Comment by ianpartridge** [comment]
I'll take a look at this.  It should "just work" (famous last words).

**Comment by ianpartridge** [comment]
Unfortunately the implementation didn't make it into the 03-24 snapshot, so this is blocked until we have migrated Kitura upwards.

**Comment by rolivieri** [comment]
@ianpartridge SwiftyJson v6.0.0 (https://github.com/IBM-Swift/SwiftyJSON/releases/tag/6.0.0) has been migrated to 4/12. Would it be possible to do this validation just using that version of SwftyJson? That way we can at least know now whether 4/12 binaries have what we need from Foundation.

**Comment by rolivieri** [comment]
@ianpartridge Hey Ian, just wondering if you have an update on this issue. Thanks!

**Comment by youming-lin** [comment]
I had another look at this issue, and due to platform differences, it's currently not possible to completely replace `LclJSONSerialization` with `NSJSONSerialization`. The main issue I encountered is `AnyObject` type casting. This code example,
```
import Foundation
var a: [String: AnyObject] = [:]
a as AnyObject
```
compiles on 06-20 toolchain for macOS, but not on the 06-20 toolchain for Ubuntu 15.10.

`NSJSONSerialzation.data` expects `AnyObject` type for the `withJSONObject` param on both platforms.

**Comment by youming-lin** [comment]
`JSONSerialization` is largely working on Linux on 0905, but I ran into an issue with serializing an array with boolean value: https://github.com/IBM-Swift/SwiftRuntime/issues/216.

**Comment by djones6** [comment]
Tracked down the commits to Foundation I think you were referring to which were blocking:
- [[SR-2694] Cannot serialize JSON object with array](https://github.com/apple/swift-corelibs-foundation/commit/f4cb7cb6ea2a9c9dedc57ea7b0069d2f7898532b)
- [Support JSONSerialization of more number types](https://github.com/apple/swift-corelibs-foundation/commit/e05884bd835f6ee22e7c0b0484326d63c31d554a)

Both are included in `swift-DEVELOPMENT-SNAPSHOT-2016-09-29-a`, but not the 3.0 stream (yet) as far as I can see.

**Comment by youming-lin** [comment]
I can confirm that all 111 tests for `SwiftyJSON`, after removing `LclJSONSerialization`, pass on the 10-04 Linux snapshot.

**Comment by collinhundley** [comment]
@youming-lin is there any reason we're still using `LclJSONSerialization` on Linux? This seems to be the case as of v1.7.

I've run into an issue where `LclJSONSerialization` fails, but I don't want to spend a lot of time tracking it down if `JSONSerialization` would be a better option.

**Comment by youming-lin** [comment]
@collinhundley We held off the change pending performance check by @djones6. @djones6 can you comment on this?

**Comment by djones6** [comment]
We did do a performance evaluation and there were several pieces of work identified that would be needed to bring `JSONSerialization` up to the same performance as we currently have in `LclJSONSerialization`.  Swift 3.1 included two PRs which brought JSONSerialization much closer to Lcl:
- https://github.com/apple/swift-corelibs-foundation/pull/718
- https://github.com/apple/swift-corelibs-foundation/pull/723

Relative performance of the two depends on what you are serializing.  Performance with Strings is equivalent, but numeric types are slower, because these are wrapped as an NSNumber and then serialized using a `CFNumberFormatter`:

![json_lcl_vs_foundation_20170209](https://user-images.githubusercontent.com/17850918/28611595-332ff6e4-71e4-11e7-8b42-a3ac10deec06.png)

@ddunn2 implemented the Int + UInt fastpath in https://github.com/apple/swift-corelibs-foundation/pull/923 (this is the green bar in the chart above), but this did not make it in time for Swift 3.1.1.  We had planned to work on a similar fastpath for Double / Float types, but held off pending an investigation of what the Linux implementation of `JSONEncoder` / `JSONDecoder` would look like.

In summary, performance of numeric types in Swift 3.1.1 is still substantially slower.  What we did in LclJSONSerialization was use `String(describing: )` for the numeric types, but whilst this is much faster, it can produce results which are inconsistent with the Darwin implementation.

**Comment by ianpartridge** [comment]
Closing as this is no longer relevant.
